db/pushdb.go

```diff
@@ -145,11 +145,36 @@ func (f *pushDatabaseOpts) AddPushServiceProviderToService(service string,
 	}
 	f.dblock.Lock()
 	defer f.dblock.Unlock()
+
+	//Service can have multiple psp in different push service type, but user might add redundant psp(different setting) with same service name. Under this situation, only one service will be used for subsribe and push in runtime.
+	//Unfortunately, no one know which psp will be used in subscribe and push later, only the very first one returned from getting psp2srv api will be used. That's a bug.
+	//So the new service psp with duplicate push type should be added. It will be verified before ADD.
+	expsps, err := f.db.GetPushServiceProvidersByService(service)
+	if err != nil {
+		return fmt.Errorf("Error querying psp with name: %v", err)
+	}
+	if len(expsps) > 0 {
+		for index := range expsps {
+			pushpsp, perr := f.db.GetPushServiceProvider(expsps[index])
+
+			if perr != nil {
+				return fmt.Errorf("Error to retrieve psp with name: %v", err)
+			}
+			//Whether the existing psp has the same push service type
+			if pushpsp.PushServiceName() == push_service_provider.PushServiceName() {
+				//The service has a psp in same push service type
+				return fmt.Errorf("Unable to add the service %s due to dulicate Push Service Provider in same type. New psp detail: %v", service, push_service_provider)
```

Comment:
Currently, you can use /addpsp to modify an existing PSP (E.g. if you inaccurately modified some of the dynamic data (VolatileData).

If the FixedData (cert, key, etc.) is identical, we already disallow that.

This is a use case we still want to support (e.g. if the default value of `bundleid` were to change, or if `sandbox` was incorrect but users were already registered, or if we add something completely different to VolatileData in the future)

```
126     // Put other things which can change in VolatileData.
127     // E.g. a bundleid can be changed by the company which manages the app.
128     if bundleid, ok := kv["bundleid"]; ok {
129         psp.VolatileData["bundleid"] = bundleid
130     } else {
131         psp.VolatileData["bundleid"] = ""
132     }
133     if sandbox, ok := kv["sandbox"]; ok {
134         if sandbox == "true" {
135             psp.VolatileData["addr"] = "gateway.sandbox.push.apple.com:2195"
136         }
137     } else {
138         if addr, ok := kv["addr"]; ok {
139             psp.VolatileData["addr"] = addr
140         } else {
141             psp.VolatileData["addr"] = "gateway.push.apple.com:2195"
142         }
143     }
144     return nil
```

Reply:
Maybe it'd make sense to forbid it only if FixedData is **different** from the currently registered PSP(s), if there are 1 or more PSPs already registered.

Reply:
Let's me explain the issue step by step in detail. For example, at the beginning, I added a apns service "myapns" with some fixed data, and point to pem key file to a path, then I update the pem file and move to a new path, I tried to update "myapns", and addpsp again. According to current logic, the second addpsp will work too, and 2 entries will exist in redis, both of them in "apns" type, have the same name "myapns", with different "fixed" data.    for the user, you have no idea which setting or "fixed data" in redis will be really used for subscribe or push when you call  service "myapns". I debug the code, and notice that only the very first setting returned from redis search will be used for push, and please notice that the pushservicetype is provided in push, so, it means you can change the pushservicetype to push to another type of psp binding to the same service "myapns", I can add a psp in gcm type to "myapns" , that's allowed according to behavior and some comments in uniqush code. 
So, the only issue i found now happens for the same service binding to to multiple pspes in same type,  under this situation. only one psp returned in redis query will be used, and the user have no idea to even know that issue.  He just keep "addpsp" to update the fixed data setting, and if fixed data are same

Reply:
Yes, consider that, we need to check fixed data under the same push service type since the psp in different push service type should be allowed to have different fixed data.
Since the fixed data is used to generate unique name of push peer. we can use Name() to compare the different fixed.
Updated pull request is launched in #201. Thanks!